### PR TITLE
Enable or Disable Multiple Create operation based on feature flag value

### PR DESCRIPTION
**Pull request by severussundar**
## Why make this change?

- Closes https://github.com/Azure/data-api-builder/issues/1951
  - PR https://github.com/Azure/data-api-builder/pull/1983, https://github.com/Azure/data-api-builder/pull/2103 add CLI options to enable or disable multiple mutation/multiple create operation through CLI. With the changes introduced in the mentioned PRs, the configuration properties successfully gets written to the config file. Also, during deserialization, the properties are read and the `MultipleMutationOptions`, `MultipleCreateOptions`, `GraphQLRuntimeOptions` objects are created accordingly.
  -  The above-mentioned PRs do not introduce any change in DAB engine behavior depending on the configuration property values.
  - This PR introduces changes to read these fields and enable/disable multiple create operation depending on whether the feature is enabled/disabled through the config file. This is achieved by introducing behavior changes in the schema generation.

## What is this change?
- This PR builds on top of a) Schema generation PR https://github.com/Azure/data-api-builder/pull/1902  b) Rename nested-create to multiple create PR https://github.com/Azure/data-api-builder/pull/2103
- When multiple create operation is disabled, 
 > i) Fields belonging to the related entities are not created in the input object type are not created.
 > ii) Many type multiple create mutation nodes (ex:  `createbooks`, `createpeople_multiple` ) are not created. 
 > iii) ReferencingField directive is not applied on relationship fields, so they continue to remain required fields for the create mutation operation.
 > iv) Entities for linking objects are not created as they are relevant only in the context of multiple create operations.

## How was this tested?

- [x] Unit Tests and Integration Tests 
- [x] Manual Tests 

**Note:** At the moment, multiple create operation is disabled in the config file generated for integration tests. This is because of the plan to merge in the Schema generation, AuthZ/N branches separately to the main branch. With just these 2 PRs, a multiple create operation will fail, hence, the disabling multiple create operation.  At the moment, tests that perform validations specific to multiple create feature enable it by i) updating the runtime object (or) ii) creating a custom config in which the operation is enabled.   

#### Related entity fields are created in the input object type
image

#### Multiple type create operation is created in addition to point create operation
image

#### Querying related entities continue to work successfully
image

#### Only fields belonging to the given entity are created in the input object type
image

#### Only fields belonging to the given entity are created in the input object type
image

### CODE REVIEW

File: src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs
Comment (ayush3797) [inline src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs:41]: Add param summary

File: src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs
Comment (ayush3797) [inline src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs:125]: nit: param summary

File: src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs
Comment (ayush3797) [inline src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs:135]: nit: split long line

File: src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs
Comment (ayush3797) [reply src/Service.GraphQLBuilder/Mutations/MutationBuilder.cs:125]: Please do the same in all other places.

File: src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs
Comment (ayush3797) [inline src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs:561]: Rather than using independent conditions: `isMultipleCreateOperationEnabled &&  DoesRelationalDBSupportMultipleCreate(databaseType)`, can we create one method which takes care of everything? Here we have two variables:

1. Database type
2. Feature-flag: set to true/false.

Obviously, I don't anticipate multiple-mutations to be supported for multiple database scenario soon. So both the above two things are known to RuntimeConfig. So maybe that might be a good place to have this method?

On a side note, I see that the RuntimeConfig.Entities are also passed on throughout all these method calls. So maybe if we can just pass the runtime config through these methods, we can take care of both the things.

File: src/Service.Tests/Configuration/ConfigurationTests.cs
Comment (ayush3797) [inline src/Service.Tests/Configuration/ConfigurationTests.cs:2068]: nit: split long line

File: src/Service.Tests/Configuration/ConfigurationTests.cs
Comment (ayush3797) [inline src/Service.Tests/Configuration/ConfigurationTests.cs:2084]: nit: split long line

File: src/Service.Tests/GraphQLBuilder/MultipleMutationBuilderTests.cs
Comment (ayush3797) [inline src/Service.Tests/GraphQLBuilder/MultipleMutationBuilderTests.cs:370]: It would be helpful to add a comment as to what the next few lines of code are doing.

File: src/Core/Services/GraphQLSchemaCreator.cs
Comment (ayush3797) [inline src/Core/Services/GraphQLSchemaCreator.cs:64]: Can all this code can be moved to `RuntimeConfig` (or any other appropriate place) and reused wherever required?

File: src/Core/Services/GraphQLSchemaCreator.cs
Comment (ayush3797) [inline src/Core/Services/GraphQLSchemaCreator.cs:269]: nit: remove blank line

Comment (ayush3797) [reply src/Core/Services/GraphQLSchemaCreator.cs:269]: I don't think this needs to be put under a conditional. The conditional logic is already there in SqlMetadataProvider to  create linking entities when multiple create is enabled: https://github.com/Azure/data-api-builder/blob/814bc34f8e6f3a79e194af62f4e71a1137e5482b/src/Core/Services/MetadataProviders/SqlMetadataProvider.cs#L761

Trying to put as less conditionals/branches as we can - will make life easier when we/other devs add support for other db types.

File: src/Service.Tests/Configuration/ConfigurationTests.cs
Comment (ayush3797) [inline src/Service.Tests/Configuration/ConfigurationTests.cs:2043]:
```suggestion
       /// <summary>
/// Test to validate that when multiple-create is disabled:
/// 1. Including a relationship field in the input for create mutation for an entity returns an exception as when multiple mutations are disabled, we don't add fields for relationships in the input type and hence users should not be able to do insertion in the related entities.
/// 2. Excluding all the relationship fields i.e.performing insertion in just the top-level entity executes successfully.
/// </summary>
```

File: src/Service.Tests/Configuration/ConfigurationTests.cs
Comment (ayush3797) [inline src/Service.Tests/Configuration/ConfigurationTests.cs:2053]: super-nit: Will be more clear to either write all of the params together or in separate lines.

File: src/Service.Tests/GraphQLBuilder/MultipleMutationBuilderTests.cs
Comment (severussundar) [reply src/Service.Tests/GraphQLBuilder/MultipleMutationBuilderTests.cs:373]: Added

File: src/Core/Services/GraphQLSchemaCreator.cs
Comment (severussundar) [reply src/Core/Services/GraphQLSchemaCreator.cs]: Moved this logic to a helper method within `RuntimeConfig`

File: src/Core/Services/GraphQLSchemaCreator.cs
Comment (severussundar) [reply src/Core/Services/GraphQLSchemaCreator.cs]: ObjectTypeDefinitionNode objects are created for linking tables only when multiple create is enabled. Adding a conditional in `GraphQLSchemaCreator` makes it easier to follow rather than letting the flow go through and a condition failing at a later stage preventing the creation of  ObjectTypeDefinitionNode objects.

File: src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs
Comment (severussundar) [reply src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs:564]: Ack. Evaluating this.

File: src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs
Comment (seantleonard) [inline src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs:1098]: nit: once `GraphQLUtils.DoesRelationalDBSupportMultipleCreate(dbType)` becomes implementation detail of `isMultipleCreateOperationEnabled` this can be modified.

File: src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs
Comment (seantleonard) [inline src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs:1057]: i understand there would be code duplication if you had a separate test was explicitly for multiple mutations. The benefits outweigh the code duplication in my opinion:
- easer to read the DataRows with way fewer parameters
- fewer conditionals in test logic that make this harder to follow
- don't have to confuse test logic that explicitly tests point mutation schema creation.

**Review by seantleonard — APPROVED** [review]
pending some suggestions about combining conditionals behind "ismultiplemutationenabled" flag. and nit about test separation. Otherwise looks good and approving so you can get those changes and merge on monday your time.

File: src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs
Comment (severussundar) [reply src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs]: I considered these factors but somehow I was leaning more towards performing validations for scenarios with feature flag on/off in the same test.

I agree with the reasoning - I've split this in two different tests. One for validating scenarios with multiple create enabled, the other with multiple create disabled.

File: src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs
Comment (severussundar) [reply src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs]: I've re-written this

File: src/Service.Tests/Configuration/ConfigurationTests.cs
Comment (severussundar) [reply src/Service.Tests/Configuration/ConfigurationTests.cs:2184]: Thanks for the suggestion.

I've added two additional tests: 
1. Point create mutation with missing `publisher_id` field with multiple create operation disabled. This should be caught by hotchocolate since a required field is missing. 
2. Point create mutation with missing `publisher_id` field with multiple create operation enabled. This should also lead to a failure. But, the error gets thrown from the database layer since there is no sufficient data to successfully execute the insert statement.  
Note: Once request validation PR gets merged, these types of requests will fail at the request validation stage (the flow will not reach till the database layer). @ayush3797, the error message that is checked in the test will have to be updated when checking-in the request validation PR.

File: src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs
Comment (severussundar) [reply src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs]: I've introduced a method called `IsMultipleCreateOperationSupportedAndEnabled()` in `RuntimeConfig` class that essentially combines the checks: a) Whether multiple create is enabled through the config file and b) Whether the database type supports multiple create operation. 

I'm still continuing to pass just the `RuntimeEntities` and `IsMultipleCreateOperationSupportedAndEnabled` parameters instead of the complete `Runtime` object as a parameter - this is because, only these 2 fields from the `RuntimeConfig` object are required. The other fields are not necessary, and hence we don't necessarily need the entire `RuntimeConfig` object.

File: src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs
Comment (seantleonard) [inline src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs:273]: naming

File: src/Service.Tests/GraphQLBuilder/Helpers/GraphQLTestHelpers.cs
Comment (seantleonard) [inline src/Service.Tests/GraphQLBuilder/Helpers/GraphQLTestHelpers.cs:46]: this hashset is now here and runtimeconfig? is the duplication necessary?

File: src/Service.Tests/Configuration/ConfigurationTests.cs
Comment (seantleonard) [inline src/Service.Tests/Configuration/ConfigurationTests.cs:2137]:
```suggestion
                // When  a create multiple operation is enabled, the "publisher_id" field will be generated as an optional field in the schema. But, when multiple create operation is disabled,
```

File: src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs
Comment (seantleonard) [inline src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs:1010]:
```suggestion
```

File: src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs
Comment (seantleonard) [inline src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs:1195]: why is this check needed when the mutation root is provided with: `IsMultipleCreateOperationSupportedAndEnabled: true)`?

File: src/Config/ObjectModel/RuntimeConfig.cs
Comment (severussundar) [reply src/Config/ObjectModel/RuntimeConfig.cs]: Good catch..

Agree that the caller of the function only needs to worry about whether the result is `true`/`false`. It does not matter if it is false because of underlying db type not supporting the operation or the operation not being enabled in the config file.

I've updated all the function and variable names to `IsMultipleCreateEnabled`

File: src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs
Comment (severussundar) [reply src/Service.Tests/GraphQLBuilder/MutationBuilderTests.cs]: Not needed, removed it.

File: src/Service.Tests/GraphQLBuilder/Helpers/GraphQLTestHelpers.cs
Comment (severussundar) [reply src/Service.Tests/GraphQLBuilder/Helpers/GraphQLTestHelpers.cs]: With the removal of conditional in `MutationBuilderTests`, this hashset is not needed anymore. Removed this.

File: src/Config/ObjectModel/RuntimeConfig.cs
Comment (severussundar) [reply src/Config/ObjectModel/RuntimeConfig.cs]: Thanks for the suggestion.

I've added a new enum name `MultipleCreateSupportingDatabaseType` - purpose is to hold all the database types that support multiple create. 

Removed the hashset from `RuntimeConfig` class.

File: src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs
Comment (severussundar) [reply src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs]: Since, https://github.com/Azure/data-api-builder/pull/2116#discussion_r1536631102 and this comment are about the same thing, we can have furhter discussion on that thread. 

Resolving this one.

File: src/Config/ObjectModel/MultipleCreateSupportingDatabaseType.cs
Comment (seantleonard) [inline src/Config/ObjectModel/MultipleCreateSupportingDatabaseType.cs:10]:
```suggestion
```

File: src/Config/ObjectModel/RuntimeConfig.cs
Comment (seantleonard) [inline src/Config/ObjectModel/RuntimeConfig.cs:467]: Can you elaborate on what is happening here? You're checking that `MultipleCreateSupportingDatabaseType` has `DatabaseType.MSSQL` defined.. which it does. This will always return true, even if the runtime config were set for DW/My/PG/Cosmos?

**Review by seantleonard — CHANGES_REQUESTED** [review]
unclear on check for MultipleCreateSupportingDatabaseType

File: src/Config/ObjectModel/RuntimeConfig.cs
Comment (seantleonard) [reply src/Config/ObjectModel/RuntimeConfig.cs]: the new implementation requires parsing an enum and iterating through MultipleCreateSupportingDatabaseType every time `IsMulitpleCreateOperationEnabled()`... seems more expensive than to just check whether configured database == mssql

File: src/Config/ObjectModel/RuntimeConfig.cs
Comment (severussundar) [reply src/Config/ObjectModel/RuntimeConfig.cs]: Thanks for spotting this...Miss on my end.

Updated to `DataSource.DatabaseType.ToString()`

File: src/Config/ObjectModel/RuntimeConfig.cs
Comment (severussundar) [reply src/Config/ObjectModel/RuntimeConfig.cs]: In the future, when the feature is supported by more database types, the enum can be updated to accommodate the new database types that support the feature.

**Review by seantleonard — APPROVED** [review]

File: src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs
Comment (severussundar) [reply src/Service.GraphQLBuilder/Mutations/CreateMutationBuilder.cs]: Changing `CreateMutationBuilder.Build()`  to accept `RuntimeObject` is purely a refactor change. IMO, we don't need the complete `RuntimeConfig` object, as only `RuntimeEntities` and `IsMultipleCreateOperationEnabled` fields are relevant for schema generation; hence, we can continue to pass them separately. 

I've created the https://github.com/Azure/data-api-builder/issues/2132 for further evaluation/discussion on this.

Since, this is purely a refactor change, this can be picked up in a follow-up PR (if we make the decision to go ahead with the refactor). This need not hold this PR from merging. 

Resolving this thread

**Pull request merged**
